### PR TITLE
Add extensions for token persistence removal feature

### PR DESCRIPTION
**Pull request by dushaniw**
### Proposed changes in this pull request

This PR is added as a part of removing token persistence efforts in https://github.com/wso2/api-manager/issues/1664.

- Provide external DAO layer plugin capability for AccessTokenDAO and TokenManagementDAO 
    Added a method to load external data access logic for AccessTokenDao and TokenManagementDao layers. With this change, APIM team hopes to plug a custom AccessTokenDAO and TokenManagementDAO implementation to support the mentioned feature.

- Add abstraction layer for Revoke flow to separate token persistence operations

    Added an abstraction layer to separate out the token dao layer access. This way, we could plug custom DAO access logic to facilitate the token persistence-related feature.
    In this pr, an interface OAuth2RevocationProcessor.java is introduced and the existing implementation parts are moved to the default class DefaultOAuth2RevocationProcessor.java

- Add abstraction layer for Refresh grant flow to separate token persistence operations
     Added a method to plug custom logic for Refresh grant handler. Reason for introducing this is because the refresh token generated in this effort is a JWT token and we need to remove token validation steps that refer stored access tokens when generating access tokens from refresh grant. This PR introduced an interface RefreshTokenGrantProcessor.java and moved the default extracted logic from RefreshGrantHandler.java to DefaultRefreshTokenGrantProcessor.java. Custom implementation can be plugged in as a service by implementing RefreshTokenGrantProcessor.java interface. If no service is registered, default DefaultRefreshTokenGrantProcessor.java is applied.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        return accessTokenDO;
+    }
+
+    private boolean validatePersistedAccessToken(RefreshTokenValidationDataDO validationBean, String clientId)
+            throws IdentityOAuth2Exception {
+
+        if (validationBean.getAccessToken() == null) {
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid Refresh Token provided for Client with " + "Client Id : " + clientId);
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:129]: Use String.format()

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    public boolean isLatestRefreshToken(OAuth2AccessTokenReqDTO tokenReq, RefreshTokenValidationDataDO validationBean,
+                                        String userStoreDomain) throws IdentityOAuth2Exception {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Evaluating refresh token. Token value: " + tokenReq.getRefreshToken() + ", Token state: " +
+                    validationBean.getRefreshTokenState());
+        }
+        if (!OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(validationBean.getRefreshTokenState())) {
+            // if refresh token is not in active state, check whether there is an access token
+            // issued with the same refresh token
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:146]: Add mutiline comments

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java (hunk 1 above)
Comment (mpmadhavig) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:129]: Do the same for other places as well

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java [hunk 3]
```diff
@@ -666,6 +670,99 @@ protected void unsetOrganizationUserResidentResolverService(
...
+            unbind = "unsetRefreshTokenGrantProcessor"
+    )
+    protected void setRefreshTokenGrantProcessor(RefreshTokenGrantProcessor refreshTokenGrantProcessor) {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Setting refresh token grant processor.");
+        }
+        OAuth2ServiceComponentHolder.getInstance().setRefreshTokenGrantProcessor(refreshTokenGrantProcessor);
+    }
+
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:687]: Let's add proper javadoc comments here. Do the same for other functions as well.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        }
+        return true;
+    }
+
+    @Override
+    public boolean isLatestRefreshToken(OAuth2AccessTokenReqDTO tokenReq, RefreshTokenValidationDataDO validationBean,
+                                        String userStoreDomain) throws IdentityOAuth2Exception {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Evaluating refresh token. Token value: " + tokenReq.getRefreshToken() + ", Token state: " +
```
Comment (mpmadhavig) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:141]: Shall we mask/remove these sensitive info such as token values.

Comment (chamathns) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:141]: Here we have tpoo use the util `IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.REFRESH_TOKEN)` and then only log the tokens

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:141]: We can check if the token whether the token is loggable and then log https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/dao/AccessTokenDAOImpl.java#L148-L157

there is one for refresh token as well
https://github.com/wso2/carbon-identity-framework/blob/master/components/identity-core/org.wso2.carbon.identity.base/src/main/java/org/wso2/carbon/identity/base/IdentityConstants.java#L340-L355

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java (hunk 2 above)
Comment (dushaniw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java]: Fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2091/commits/a79ee2e471e71fa815a54c003704a15f56970f17

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/RefreshTokenGrantProcessor.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.wso2.carbon.identity.oauth2.model.AccessTokenDO;
+import org.wso2.carbon.identity.oauth2.model.RefreshTokenValidationDataDO;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+
+/**
+ * Intermediate processor for handling refresh token persistence logic.
+ */
+public interface RefreshTokenGrantProcessor {
+
+    RefreshTokenValidationDataDO validateRefreshToken(OAuthTokenReqMessageContext tokenReqMessageContext)
```
[the remark this answers is not in the document]
Comment (dushaniw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/RefreshTokenGrantProcessor.java:39]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2091/commits/d3dfa2211975383ff488e13e4313809db84573cf

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java (hunk 3 above)
Comment (dushaniw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:697]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2091/commits/d3dfa2211975383ff488e13e4313809db84573cf

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java (hunk 4 above)
Comment (dushaniw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2091/commits/ff7a124dc1d0d41040fde1134ad8b524f98a8329

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+                        validationBean.getRefreshTokenState()));
+            }
+        }
+        if (!OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE.equals(validationBean.getRefreshTokenState())) {
+            /* if refresh token is not in active state, check whether there is an access token issued with the same
+             * refresh token.
+             */
+            List<AccessTokenDO> accessTokenBeans = getAccessTokenBeans(tokenReq, validationBean, userStoreDomain);
+            for (AccessTokenDO token : accessTokenBeans) {
+                if (tokenReq.getRefreshToken().equals(token.getRefreshToken())
```
Comment (chamathns) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java:156]: is this null-safe?

Comment (dushaniw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java]: This is the refresh token sent in the request for refresh grant itself. Couldnt find a place in the inbound code validating the existence of the parameter hence adding a null check.

Comment (dushaniw) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/tokenprocessor/DefaultRefreshTokenGrantProcessor.java]: fixed with https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2091/commits/be0223e8df1b84d536c22ea214ef247d84995dcc

**Review by chamathns — APPROVED** [review]

**Pull request merged**